kong/db/dao/plugins/external.lua

```diff
@@ -0,0 +1,642 @@
+      plugin_info.Name, _plugin_infos[plugin_info.Name].server_def.name, server_def.name))
+    return
+  end
+
+  _plugin_infos[plugin_info.Name] = {
+    server_def = server_def,
+    --rpc = server_def.rpc,
+    name = plugin_info.Name,
+    PRIORITY = plugin_info.Priority,
+    VERSION = plugin_info.Version,
+    schema = plugin_info.Schema,
+    phases = plugin_info.Phases,
+  }
+end
+
+local function ask_info(server_def)
+  if not server_def.query_command then
+    kong.log.info(string.format("No info query for %s", server_def.name))
+    return
+  end
+
+  local fd, err = io.popen(server_def.query_command)
+  if not fd then
+    local msg = string.format("loading plugins info from [%s]:\n", server_def.name)
+    kong.log.err(msg, err)
+    return
+  end
+
+  local infos_dump = fd:read("*a")
+  fd:close()
+  local infos = cjson_decode(infos_dump)
+  if type(infos) ~= "table" then
+    error(string.format("Not a plugin info table: \n%s\n%s",
+        server_def.query_command, infos_dump))
+    return
+  end
+
+  for _, plugin_info in ipairs(infos) do
+    register_plugin_info(server_def, plugin_info)
+  end
+end
+
+local function load_all_infos()
+  if not _plugin_infos then
+    _plugin_infos = {}
+
+    for _, server_def in ipairs(get_server_defs()) do
+      ask_info(server_def)
+    end
+  end
+
+  return _plugin_infos
+end
+
+
+local loaded_plugins = {}
+
+local function get_plugin(plugin_name)
+  if not loaded_plugins[plugin_name] then
+    local plugin = load_all_infos()[plugin_name]
+    loaded_plugins[plugin_name] = build_phases(plugin)
+  end
+
+  return loaded_plugins[plugin_name]
+end
+
+function external_plugins.load_plugin(plugin_name)
+  local plugin = get_plugin(plugin_name)
+  if plugin and plugin.PRIORITY then
+    return true, plugin
+  end
+
+  return false, "no plugin found"
+end
+
+function external_plugins.load_schema(plugin_name)
+  local plugin = get_plugin(plugin_name)
+  if plugin and plugin.PRIORITY then
+    return true, plugin.schema
+  end
+
+  return false, "no plugin found"
+end
+
+
+--[[
+
+Process management
+
+Servers that specify an `.exec` field are launched and managed by Kong.
+This is an attempt to duplicate the smallest reasonable subset of systemd.
+
+Each process specifies executable, arguments and environment.
+Stdout and stderr are joined and logged, if it dies, Kong logs the event
+and respawns the server.
+
+--]]
+
+local function grab_logs(proc, name)
+  local prefix = string.format("[%s:%d] ", name, proc:pid())
+
+  while true do
+    local data, err, partial = proc:stdout_read_line()
+    local line = data or partial
+    if line and line ~= "" then
+      raw_log(ngx_INFO, prefix .. line)
+    end
+
+    if not data and err == "closed" then
+      return
+    end
+  end
+end
+
+local function handle_server(server_def)
+  if not server_def.socket then
+    -- no error, just ignore
+    return
+  end
+
+  if server_def.start_command then
+    ngx_timer_at(0, function(premature)
```

Comment:
I suggest taking this function out to a separate function, adding the `server_def` as parameter:

```
function plugin_server_timer(premature, server_def)
  if premature then
  etc etc
```

Then the preambles are short enough that `handle_server` is not needed any more and can be put in the loop inside `external_plugins.manage_servers`